### PR TITLE
Feature/refactor s3 source

### PR DESCRIPTION
**Pull request by jayfjin**
JIRA - https://issues.cask.co/browse/CDAP-3152
BUILD - http://builds.cask.co/browse/CDAP-DUT2402-1

Creates a S3BatchSource that extends FileBatchSource. Also refactored TPFS Avro/Parquet Sink.

Old PR got nuked by a merge and had an LGTM minus the new .rst file I added (s3.rst).

### CODE REVIEW

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/source/S3BatchSource.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+import com.google.common.collect.ImmutableMap;
+import com.google.gson.Gson;
+
+import javax.annotation.Nullable;
+
+/**
+ * A {@link BatchSource} that reads from Amazon S3.
+ */
+@Plugin(type = "source")
+@Name("S3BatchSource")
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/source/S3BatchSource.java]: S3

File: cdap-docs/application-templates/source/etl/templates/sources/batch/s3.rst [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+subclass of FileInputFormat. Defaults to TextInputFormat.
+
+**maxSplitSize:** Maximum split-size for each mapper in the MapReduce Job. Defaults to 128MB.
+
+.. rubric:: Example
+
+::
+
+  {
+  "name": "FileBatchSource",
```
Comment (albertshau) [inline cdap-docs/application-templates/source/etl/templates/sources/batch/s3.rst]: indentation

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/source/S3BatchSource.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+@Description("Batch source for Amazon S3")
+public class S3BatchSource extends FileBatchSource {
+  private static final String ACCESS_ID_DESCRIPTION = "Access ID of Amazon S3 instance to connect to.";
```
Comment (awholegunch) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/source/S3BatchSource.java]: the Amazon S3 instance

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/source/S3BatchSource.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  private static final String ACCESS_KEY_DESCRIPTION = "Access Key of Amazon S3 instance to connect to.";
```
Comment (awholegunch) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/source/S3BatchSource.java]: the Amazon S3 instance

File: cdap-docs/application-templates/source/etl/templates/sources/batch/s3.rst [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+      "maxSplitSize": "1048576"
+    }
+  }
+
+This example connects to Amazon S3 and reads in files found in the specified directory while
+using the stateful Timefilter, which ensures that each file is read only once. The Timefilter
+requires that files be named with either the convention "yy-MM-dd-HH..." (S3) or "...'.'yy-MM-dd-HH..."
+(Cloudfront). The stateful metadata is stored in a table named 'timeTable'. With the maxSplitSize
+set to 1MB, if the total size of the files being read is larger than 1MB, CDAP will
+configure Hadoop to use more than one mapper.
```
Comment (albertshau) [inline cdap-docs/application-templates/source/etl/templates/sources/batch/s3.rst]: does this mean 1 mapper per 1mb? more than one mapper is a bit ambiguous

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/source/FileBatchSource.java [hunk 6]
```diff
@@ -71,8 +72,15 @@
     Schema.Field.of("ts", Schema.of(Schema.Type.LONG)),
     Schema.Field.of("body", Schema.of(Schema.Type.STRING))
   );
-
-  private static final String REGEX_DESCRIPTION = "Regex to filter out filenames in the path. " +
+  protected static final String MAX_SPLIT_SIZE_DESCRIPTION = "Maximum split size for each mapper in the MapReduce " +
```
Comment (awholegunch) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/source/FileBatchSource.java]: split-size

**Comment by albertshau** [comment]
content looks good, will leave style comments to jj

File: cdap-docs/application-templates/source/etl/templates/sources/batch/s3.rst (hunk 2 above)
Comment (awholegunch) [inline cdap-docs/application-templates/source/etl/templates/sources/batch/s3.rst]: good spotting.

**Comment by awholegunch** [comment]
Docs LGTM

**Comment by jayfjin** [comment]
waiting for build to pass, then merging: http://builds.cask.co/browse/CDAP-DUT2402-6

**Pull request merged**